bindings/java/lib/native/src/lib.rs

```diff
@@ -150,77 +178,108 @@ pub unsafe extern "system" fn Java_org_iota_api_NativeApi_listen(
-                let mut events_list: Vec<WalletEventType> = Vec::with_capacity(string_count.try_into().unwrap());
-
-                for i in 0..string_count {
-                    let java_str: String = env
-                        .get_string(env.get_object_array_element(events, i).unwrap().into())
-                        .expect("Couldn't get java string!")
-                        .into();
-                    events_list.push(WalletEventType::try_from(&*java_str).expect("Unknon WalletEventType"));
-                }
-
-                // Allocate callback globaly to not lose the reference
-                let global_obj = env
-                    .new_global_ref(JObject::from_raw(callback))
-                    .expect("Failed to make Global ref");
-
-                crate::block_on(message_handler.listen(events_list, move |e| {
-                    let ev_ser = serde_json::to_string(&e).expect("Failed to serialise event");
-
-                    // Grab env from VM
-                    let vm_guard = VM.lock().unwrap();
-                    let env = {
-                        let vm = vm_guard.as_ref().unwrap();
-                        let mut env = vm.get_env();
-                        if env.is_err() {
-                            env = vm.attach_current_thread().map(|e| *e);
-                        }
-                        env.expect("failed to get env")
-                    };
-
-                    let event =
-                        JValue::Object(*env.new_string(ev_ser).expect("Couldn't create event as string!")).to_jni();
-                    let cb = JValue::Object(global_obj.as_obj()).to_jni();
-
-                    env.call_static_method_unchecked(
-                        "org/iota/api/NativeApi",
-                        METHOD_CACHE.lock().unwrap().unwrap(),
-                        ReturnType::Object,
-                        &[event, cb],
-                    )
-                    .expect("Failed to call handleCallback");
-                }));
-
-                let output = env
-                    .new_string("{\"type\": \"success\", \"payload\": \"success\"}")
-                    .expect("Couldn't create java string!");
-
-                output.into_raw()
-            }
-            None => {
-                env.throw_new("java/lang/NullPointerException", "Wallet not initialised.")
-                    .unwrap();
-                std::ptr::null_mut()
+                crate::block_on(message_handler.listen(events_list, move |e| event_handle(global_obj.clone(), e)))
             }
+            _ => throw_nullpointer(&env, "Wallet not initialised."),
         },
-        Err(err) => {
-            env.throw_new("java/lang/Exception", err.to_string()).unwrap();
-            std::ptr::null_mut()
+        Err(err) => throw_exception(&env, err.to_string()),
+    };
+
+    // exceptions return early, let the client now we registered successfully
```

Comment:
Don't understand this part `let the client now we registered`

Reply:
now -> know. typo :D